frail/ast.py

```diff
@@ -2,7 +2,7 @@
 from typing import Callable, Dict, Union
 
 
-@dataclass(eq=True, frozen=True)
```

Comment:
Why did you need to unfreeze these? Are you modifying the nodes node-referential properties rather than creating new ones? Or is it just changing references?

Reply:
Yes, I'm not creating new nodes and am changing the nodes themselves. For example, in the strength rewrite, a node has arg0_index as a MulOp that is now been replaced by a counter val, this arg0_index is changed to be the counter index instead of the original multiply one.

Reply:
That should work for the strength reduction rewrite because we're replacing equivalent nodes. However, for nested counter, I'm surprised you didn't run into issues. It's fine for now if no problems

In general, I prefer the dag to be immutable and then make a new dag after each rewrite. This makes debugging easier as you can look at the dag after each transformation and compare what happened. But not worth doing significant rewrites currently since your approach works.

Reply:
Sounds good, can definitely add that as a todo for a future PR.